Rework the handler code generation

jbachorik opened the pull request:
Do not rely on generic deserialization methods accessed via method handles for reading, and most importantly skipping values.

Generate in-line bytecode to fully deserialize or skip full type instances.

Fixes #6 

---
### (Naive) Benchmarks
What I mean by 'naive' is that this is just a single particular workload - although it does exercise event filtering, proper deserialization from constant pool and also direct read from the recording.

The test will iterate over all `ExecutionSample` events, count the number of events and sums all thread ids and the depths of the event stacktraces.

Running against 1.79GiB JFR file generated by JFR proper.

The test machine is MacPro M1, Java 21.0.3

#### Jafar
```
jaroslav.bachorik@COMP-XLQW7Y30T0 jafar % time java -cp demo/build/libs/demo-all.jar io.jafar.demo.Main jafar demo/src/test/resources/test-jfr.jfr
Total events: 176592
Sum of thread ids: 6838372957
java -cp demo/build/libs/demo-all.jar io.jafar.demo.Main jafar   12.60s user 1.54s system 701% cpu 2.014 total
```
#### JFR CLI
```
jaroslav.bachorik@COMP-XLQW7Y30T0 jafar % time java -cp demo/build/libs/demo-all.jar io.jafar.demo.Main jfr demo/src/test/resources/test-jfr.jfr 
Total events: 176592
Sum of thread ids: 6838372957
java -cp demo/build/libs/demo-all.jar io.jafar.demo.Main jfr   75.37s user 3.52s system 329% cpu 23.925 total
```

#### JFR Streaming
```
jaroslav.bachorik@COMP-XLQW7Y30T0 jafar % time  java -jar demo/build/libs/demo-all.jar jfr-stream ~/Downloads/test-jfr.jfr
Total events: 176592
Sum of thread ids: 6838372957
java -jar demo/build/libs/demo-all.jar jfr-stream ~/Downloads/test-jfr.jfr  93.50s user 3.64s system 394% cpu 24.595 total
```

#### JMC Parser
```
jaroslav.bachorik@COMP-XLQW7Y30T0 jafar % time java -cp demo/build/libs/demo-all.jar io.jafar.demo.Main jmc demo/src/test/resources/test-jfr.jfr
Total events: 176592
Sum of thread ids: 6838372957
java -cp demo/build/libs/demo-all.jar io.jafar.demo.Main jmc   69.54s user 6.55s system 627% cpu 12.118 total
```

#### Async-profiler Parser
For this I had to slightly hack the code in the async-profiler repo (https://github.com/jbachorik/async-profiler/blob/jb/jafar_test/src/converter/one/jfr/Main.java) to get the exact same functionality so I could compare the performance. You can checkout the fork/branch, build with `make all` and then run the benchmark.

```
jaroslav.bachorik@COMP-XLQW7Y30T0 async-profiler % time java -cp /Users/jaroslav.bachorik/opensource/async-profiler/build/jar/jfr-converter.jar one.jfr.Main <path_to_jafar>/demo/src/test/resources/test-jfr.jfr
Total events: 185881
Sum of thread ids: 6920945493
java -cp  one.jfr.Main <path_to_jafar>/demo/src/test/test-jfr.jfr  7.67s user 456.57s system 100% cpu 7:39.94 total
```

The results are quite bad for the async-profiler parser. This seems to be because it is optimized for the JFR files generated by async-profiler, where it beats all other parsers (you can check yourself by using `test-ap.jfr` file from the demo test resources). What is more worrying is that the number of reported events is significantly higher than the one reported by all the other parsers.

--- 
Looking at the relative speedup we can see that Jafar parser is around 5 times more efficient than the `jfr` tool parser or even the `JMC` parser.

| Parser           | Time (User + Sys) | Time % | Wall Time | Wall Time % |
|-------------|--------------------|--------|-----------|--------------|
| JFR              |  78.89                      | 100       | 23.9          | 100                 |
| JFR Stream |  97.14                      |  123.      |  24.595.   | 102                 |
| JMC            |  76                           | 96        | 12.118        |  51                   |
| Jafar           |   14.1                         | 18         |  2.014       |    9                  |
| AP               |   464.24                   | 588      | 459.94     |  1924              |

egahlin commented:
It would be interesting to get the numbers on more normal sized files, i.e. 50 MB, and with the JFR streaming parser, which can skip events on file and pool objects. 

```
var es  = EventStream.openFile(...);
es.setReuse(true);
es.setOrdered(false);
es.onEvent("jdk.ExecutionSample", e -> {
});
es.start();
```

Pooling is typically not a win for larger recordings, but may avoid heap expansion on smaller recordings and reduce GCs on latency sensitive applications.

egahlin commented:
> I have added the jfr-stream option to the demo app and added the results.
> 
> I will collect ~50MiB recording and add also the table for that. Preliminary, for a non-representative ~20MiB JFR recording I still see ~2x improvement with Jafar parser but the actual runtime is so short (under 0.5s) that it does not really make difference.

Interesting, I would guess that object pooling and that the recording is not representative are the main reasons it is slower than RecordingFile. Skipping events should be a huge win if execution samples only make up 5% of the events in a file.

When I did the initial JMC parser in 2009, I experimented with a bytecode-generated class for every event type, but it made opening files slower in JMC due to compilation, and the GUI layer had to box the values anyway because writing accumulator logic (sum, average, etc.) and formatting for different primitive types became too cumbersome. With method handles and Hotspot (with an interpreter), it should work better. 

Anyway, good results!

jbachorik commented:
> > I have added the jfr-stream option to the demo app and added the results.
> > I will collect ~50MiB recording and add also the table for that. Preliminary, for a non-representative ~20MiB JFR recording I still see ~2x improvement with Jafar parser but the actual runtime is so short (under 0.5s) that it does not really make difference.
> 
> Interesting, I would guess that object pooling and that the recording is not representative are the main reasons it is slower than RecordingFile. Skipping events should be a huge win if execution samples only make up 5% of the events in a file.
> 
> When I did the initial JMC parser in 2009, I experimented with a bytecode-generated class for every event type, but it made opening files slower in JMC due to compilation, and the GUI layer had to box the values anyway because writing accumulator logic (sum, average, etc.) and formatting for different primitive types became too cumbersome. With method handles and Hotspot (with an interpreter), it should work better.
> 
> Anyway, good results!

Thanks! This parser is not very suitable for the dynamic UI of JMC. It is really good for situations you know that you are going to need just a few event types and will be using just a few attributes from them. That's what allows the optimizations.

egahlin commented on demo/src/main/java/io/jafar/demo/Main.java:
```diff
@@ -71,16 +76,33 @@ private static void runWithJafar(String[] args, LongAccumulator sum, AtomicInteg
+            cnt.incrementAndGet();
+        });
+        es.start();
+        try {
```
If start() is used, the parser will run in the same thread until there is no more data. No need to awaitTermination().
